eventlet 0.36.0 incompatible with flask-socketio running under gunicorn

Haven't figured out a minimal test case for this yet, but our CI pipelines for https://github.com/roedoejet/g2p/ have all stopped working because of the 0.36.0 release.  We are running a background flask-socketio server with gunicorn and eventlet, and trying to access it via Playwright, which hangs forever when trying to connect to the socket.

The documentation seems to indicate that we should not use eventlet anymore, so we'll do that :) but just to note that the most recent eventlet seems to be particularly ill-behaved.

Thank you! At one point I already recreated our API + web app using FastAPI, but we didn't end up switching because Flask is fairly embedded in our codebase - so if Quart has the same API it seems like the migration would be simpler.

We don't particularly care about asyncio (though I agree that it is better), we just need something that supports SocketIO, but it seems that to do so with Flask in production requires eventlet, gevent, or something else, which from what I understand are all somewhat of a compromise.

To reproduce the error, it's not all browsers that seem to trigger it - Safari definitely does, while Chrome on Linux seems to work okay, yet somehow Playwright hangs in testing.  Full sequence to get the error is:

```
git clone https://github.com/roedoejet/g2p.git
pip install g2p
gunicorn --worker-class eventlet -w 1 g2p.app:APP --no-sendfile --bind 0.0.0.0:5000
python g2p/g2p/tests/test_studio.py  # in another window obiously
```

Actually it seems that the error is being triggered by the websocket upgrade request:

```
[2024-03-26 08:45:40 -0400] [16229] [ERROR] Error handling request /socket.io/?EIO=4&transport=websocket&sid=62vsmkRvQIZ13WscAAAU
Traceback (most recent call last):
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/gunicorn/workers/base_async.py", line 55, in handle
    self.handle_request(listener_name, req, client, addr)
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/gunicorn/workers/base_async.py", line 108, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/flask/app.py", line 2552, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/flask_socketio/__init__.py", line 43, in __call__
    return super(_SocketIOMiddleware, self).__call__(environ,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/engineio/middleware.py", line 63, in __call__
    return self.engineio_app.handle_request(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/socketio/server.py", line 430, in handle_request
    return self.eio.handle_request(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/engineio/server.py", line 274, in handle_request
    packets = socket.handle_get_request(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/engineio/socket.py", line 90, in handle_get_request
    return getattr(self, '_upgrade_' + transport)(environ,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/engineio/socket.py", line 146, in _upgrade_websocket
    return ws(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/engineio/async_drivers/eventlet.py", line 44, in __call__
    return super().__call__(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dhd/work/ReadAlongs/venv/lib/python3.11/site-packages/eventlet/websocket.py", line 137, in __call__
    environ['eventlet.set_idle']()
    ~~~~~~~^^^^^^^^^^^^^^^^^^^^^
KeyError: 'eventlet.set_idle'
```

--- a/eventlet/websocket.py
+++ b/eventlet/websocket.py
@@ -132,10 +132,11 @@ class WebSocketWSGI:
                            [('Connection', 'close'), ] + headers)
             return [body]
 
-        # We're ready to switch protocols; flag the connection
-        # as idle to play well with a graceful stop
-        environ['eventlet.set_idle']()
-
+        # We're ready to switch protocols; if running under Eventlet
+        # (this is not always the case) then flag the connection as
+        # idle to play well with a graceful stop
+        if 'eventlet.set_idle' in environ:
+            environ['eventlet.set_idle']()
         try:
             self.handler(ws)
         except OSError as e:

